### DISCUSSION TITLE
[Run Issue]: Visualization Freezes During Cell Simulation

### DISCUSSION THREAD

**Discussion by ghost** [topas-nbio/TOPAS-nBio#46] [Run-Problem] [answered]
### Contact Details

cecilianovawijaya01@gmail.com

### What was the issue?

Hello! 

I am trying to simulate a cell with a DNA structure inside. I want to activate Qt and OpenGL to visualize the geometry I’ve set (to check if it’s correct). However, when I run it, the Qt screen doesn’t show anything and gradually freezes. I even tried using another computer with 128 threads, but it still doesn’t work.

Here is my code:
```
# ----- Material -----

# Cytoplasm
sv:Ma/Cytoplasm/Components	= 4 "Hydrogen" "Carbon" "Nitrogen" "Oxygen"
uv:Ma/Cytoplasm/Fractions	= 4 0.1025 0.1225 0.0425 0.7325
d:Ma/Cytoplasm/Density		= 1.0 g/cm3
s:Ma/Cytoplasm/DefaultColor	= "blue"

# Nucleoplasm
sv:Ma/Nucleoplasm/Components	= 6 "Hydrogen" "Carbon" "Nitrogen" "Oxygen" "Sulfur" "Potassium"
uv:Ma/Nucleoplasm/Fractions	= 6 0.106 0.09 0.032 0.742 0.004 0.026
d:Ma/Nucleoplasm/Density	= 1.0 g/cm3
s:Ma/Nucleoplasm/DefaultColor	= "red"

s:Ma/G4_HistoneMaterial/CloneFromMaterial   = "G4_WATER"
d:Ma/G4_HistoneMaterial/CloneWithDensity    = 1.407 g/cm3
s:Ma/G4_BaseMaterial/CloneFromMaterial      = "G4_WATER"
d:Ma/G4_BaseMaterial/CloneWithDensity       = 1.0 g/cm3
s:Ma/G4_BackboneMaterial/CloneFromMaterial  = "G4_WATER"
d:Ma/G4_BackboneMaterial/CloneWithDensity   = 1.0 g/cm3
s:Ma/G4_WATER_MODIFIED/CloneFromMaterial    = "G4_WATER"
d:Ma/G4_WATER_MODIFIED/CloneWithDensity     = 1.0 g/cm3

# ----- Geometry -----

# World
d:Ge/World/HLX		= 30 um
d:Ge/World/HLY		= 30 um
d:Ge/World/HLZ		= 30 um
s:Ge/World/Material	= "G4_WATER"
s:Ge/World/Color	= "Aqua"
s:Ge/World/DrawingStyle	= "WireFrame"

# Target
s:Ge/Target/Type	= "G4Orb"
s:Ge/Target/Parent	= "World"
s:Ge/Target/Material	= "G4_WATER"
d:Ge/Target/R		= 10.1 um
s:Ge/Target/Color	= "aqua"
s:Ge/Target/DrawingStyle= "WireFrame"

# Cell
s:Ge/Cell/Type		= "TsSphericalCellSphericalNP"
s:Ge/Cell/Material	= "Cytoplasm" # "G4_WATER"
s:Ge/Cell/Parent	= "Target"
d:Ge/Cell/CellRadius	= 10 um
s:Ge/Cell/Color		= "green"
s:Ge/Cell/DrawingStyle	= "WireFrame"

# Nucleus
s:Ge/Cell/Nucleus/Type			= "TsNucleus"
s:Ge/Cell/Nucleus/Parent		= "Cell"
s:Ge/Cell/Nucleus/Material		= "Nucleoplasm" # "G4_WATER"
s:Ge/Cell/Nucleus/Color			= "White"
s:Ge/Cell/Nucleus/DNAModel		= "Sphere"
d:Ge/Cell/Nucleus/NucleusRadius		= 6.9 um
i:Ge/Cell/Nucleus/HilbertCurveLayer	= 2
i:Ge/Cell/Nucleus/HilbertCurve3DRepeat	= 36
s:Ge/Cell/Nucleus/HilbertCurveFileName	= "supportFiles/HilbertPoints_iteration1.dat"
s:Ge/Cell/Nucleus/Voxel/Material	= "G4_WATER"
b:Ge/Cell/Nucleus/ShowDNAVoxels		= "False"
s:Ge/Cell/Nucleus/Histone/Material	= "G4_HistoneMaterial"
b:Ge/Cell/Nucleus/OnlyBuildOneHistone	= "False"
b:Ge/Cell/Nucleus/HistoneAsScavenger	= "True"
b:Ge/Cell/Nucleus/AddBases		= "True"
s:Ge/Cell/Nucleus/Base1/Material	= "G4_BaseMaterial"
s:Ge/Cell/Nucleus/Base2/Material	= "G4_BaseMaterial"
b:Ge/Cell/Nucleus/AddBackbones		= "True"
s:Ge/Cell/Nucleus/Backbone1/Material	= "G4_BackboneMaterial"
s:Ge/Cell/Nucleus/Backbone2/Material	= "G4_BackboneMaterial"
b:Ge/Cell/Nucleus/AddHydrationShell	= "True"
s:Ge/Cell/Nucleus/ChromatinFiber/Material	= "G4_WATER"
b:Ge/Cell/Nucleus/ShowChromatinCylinders	= "False"
s:Ge/Cell/Nucleus/HydrationShell1/Material	= "G4_WATER_MODIFIED"
s:Ge/Cell/Nucleus/HydrationShell2/Material	= "G4_WATER_MODIFIED"


# Nanoparticle
i:Ge/Cell/Nanoparticle/NumberOfNanoparticles	= 1
i:Ge/Cell/Nanoparticle/NumberOfNanoparticlesAtNucleus	= 1
i:Ge/Cell/Nanoparticle/NumberOfNanoparticlesAtMitochondria	= 1
d:Ge/Cell/Nanoparticle/r			= 50 nm
s:Ge/Cell/Nanoparticle/Material			= "G4_Au"
s:Ge/Cell/Nanoparticle/Color			= "yellow"
s:Ge/Cell/Nanoparticle/DrawingStyle		= "solid"

# ----- Graphics -----

s:Gr/ViewA/Type			= "OpenGL"
b:Gr/ViewA/IncludeAxes		= "True" 
s:Gr/ViewA/AxesComponent	= "World" 
d:Gr/ViewA/AxesSize		= 2.0 um

Ts/UseQt = "True"

# ----- Program Control -----

i:Ts/ShowHistoryCountAtInterval		= 10
i:Ts/NumberOfThreads			= -5
b:Ts/PauseBeforeQuit			= "True" 
b:Ts/ShowCPUTime			= "True"
b:Ge/CheckForOverlaps			= "False"
b:Ge/CheckInsideEnvelopesForOverlaps	= "False"
```

Here is the output:
![Selection_055](https://github.com/user-attachments/assets/ae8dddff-5ccd-4235-ab10-9518be47a8e5)

When I turn off the visualization and Qt, it runs perfectly:
![Selection_056](https://github.com/user-attachments/assets/d6aa27c9-803a-40b0-ba27-4f2af6455588)

Is there anything wrong with my code? Should I turn something off? Or could it simply mean that my computer doesn’t have enough power to simulate this graphic setup?

Thank you for your attention!

### TOPAS-nBio category

Geometry

### Version

3.0

### What Operative System (OS) are you running on?

Linux (Debian/Ubuntu)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Attestation

- [X] I testify that I have fully read the bug reporting guideline beforehand.

**Comment by tiagoaaazevedo1** [comment, marked as answer]
Hello! 

I ran your code on my MacBook, and it also froze. I tried running your example without the "TsNucleus" and it runs without any problem. From my experience, the "TsNucleus" geometry is very detailed and computationally intensive, and I never could visualize it. I think the best proxy for that geometry is the "TsSolenoidFiber" which is relatively easy to run, and It could help with the geometrical setup.

Hope it helps! :)
Best regards,
Tiago A.

**Reply by ghost** [reply]
Hello Tiago,

Thank you for answering my question! Ah, I see. I am glad to hear that the issue has nothing to do with the code itself.

When you recommended using `TsSolenoidFiber` for the DNA structure, does this mean I should define two separate geometries (one for the nucleus and one for the DNA strand)? My understanding is that `TsNucleus` models a cell nucleus as a sphere uniformly filled with DNA, whereas `TsSolenoidFiber` represents a chromatin fiber model based on solenoid geometry.

Thank you very much for your help and guidance.

Best,
Cecilia

**Reply by tiagoaaazevedo1** [reply]
In this case, I suggested using the chromatin fiber only to precheck the geometry. From what I understand, the TsNucleus model is built upon the TsSolenoidFiber model by arranging multiple solenoid fibers in space. So to fully simulate the DNA damage in the nucleus, the TsNucleus is a more realistic representation, although that comes with a computational cost, especially in the visualization.

I hope this helps! Let me know if you have any more questions :)

Tiago A.

**Reply by ghost** [reply]
Hello Tiago,

I see. I will use the TsNucleus for my simulation but I will try to use TsSolenoidFiber to simulate and check the geometry.

Thank you for your advice! 

Best,
Cecilia
